manta-pay/src/crypto/hash/poseidon/parameter_generation/preprocessing.rs

```diff
@@ -0,0 +1,100 @@
+// Copyright 2019-2022 Manta Network.
+// This file is part of manta-rs.
+//
+// manta-rs is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// manta-rs is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with manta-rs.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Preprocess constants for optimized poseidon hash
+//! acknowledgement: adapted from FileCoin Project: <https://github.com/filecoin-project/neptune/blob/master/src/preprocessing.rs>
+
+use super::{matrix::vec_add, mds::MdsMatrices};
+use crate::crypto::hash::ParamField;
+use alloc::vec::Vec;
+
+/// Compresses constants by pushing them back through linear layers and through the identity components of partial layers.
+/// As a result, constants need only be added after each S-box.
+pub fn compress_round_constants<F>(
+    width: usize,
+    full_rounds: usize,
+    partial_rounds: usize,
+    round_constants: &[F],
+    mds_matrices: &MdsMatrices<F>,
+) -> Vec<F>
+where
+    F: ParamField + Copy,
+{
+    let inverse_matrix = &mds_matrices.m_inv;
+
+    let mut res: Vec<F> = Vec::new();
+
+    let round_keys = |r: usize| &round_constants[r * width..(r + 1) * width];
+
+    // This is half full-rounds.
+    let half_full_rounds = full_rounds / 2;
+
+    // First round constants are unchanged.
+    res.extend(round_keys(0));
+
+    // Post S-box adds for the first set of full rounds should be 'inverted' from next round.
+    // The final round is skipped when fully preprocessing because that value must be obtained from the result of preprocessing the partial rounds.
+    let end = half_full_rounds - 1;
+    for i in 0..end {
+        let next_round = round_keys(i + 1);
+        let inverted = inverse_matrix.right_apply(next_round);
+        res.extend(inverted);
+    }
+
+    // The plan:

```

Comment:
Still a plan? If not remove this

Reply:
I changed the name from "the plan" to "high-level workflow description". Since the computation here is complex, we need this description as an intuitive explanation of what is happening.